From 9b2540f1db53a3639cf2ee8c58a1323700a501b0 Mon Sep 17 00:00:00 2001
From: Laibin Qiu <qiulaibin@huawei.com>
Date: Mon, 15 Nov 2021 19:35:53 +0800
Subject: [PATCH] block: fix memory leak for mq shared sbitmap
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 9b2540f1db53a3639cf2ee8c58a1323700a501b0


hulk inclusion
category: bugfix
bugzilla: 182674 https://gitee.com/openeuler/kernel/issues/I4DDEL

---------------------------

If the blk_mq_sched_alloc_tags() -> blk_mq_alloc_rqs() call fails,
then we call blk_mq_free_rq_map(). But if BLK_MQ_F_TAG_HCTX_SHARED
has been set to hctx->flags, tags->bitmap_tags and
tags->breserved_tags will not be released and cause a memory leak.

Fixes: 31044c2e5b8de ("blk-mq-sched: Fix blk_mq_sched_alloc_tags() error
handling")
Signed-off-by: Laibin Qiu <qiulaibin@huawei.com>
Reviewed-by: Jason Yan <yanaijie@huawei.com>

Signed-off-by: Chen Jun <chenjun102@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 block/blk-mq-sched.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/blk-mq-sched.c b/block/blk-mq-sched.c
index d495c5fe5e6d..090aa0416443 100644
--- a/block/blk-mq-sched.c
+++ b/block/blk-mq-sched.c
@@ -521,7 +521,7 @@ static int blk_mq_sched_alloc_tags(struct request_queue *q,
 
 	ret = blk_mq_alloc_rqs(set, hctx->sched_tags, hctx_idx, q->nr_requests);
 	if (ret) {
-		blk_mq_free_rq_map(hctx->sched_tags, set->flags);
+		blk_mq_free_rq_map(hctx->sched_tags, flags);
 		hctx->sched_tags = NULL;
 	}
 
-- 
2.26.2

